File: JAVA/java/src/PalindromeLinkedList.java
// Rename ListNode class to MyListNode
 class MyListNode {
    int val;
    MyListNode next;
    MyListNode(int x) { val = x; }
}

public class PalindromeLinkedList {
    public boolean isPalindrome(MyListNode head) {
        MyListNode fast = head, slow = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        if (fast != null) { // odd nodes: let right half smaller
            slow = slow.next;
        }
        slow = reverse(slow);
        fast = head;

        while (slow != null) {
            if (fast.val != slow.val) {
                return false;
            }
            fast = fast.next;
            slow = slow.next;
        }
        return true;
    }

    public MyListNode reverse(MyListNode head) {
        MyListNode prev = null;
        while (head != null) {
            MyListNode next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }

    public static void main(String[] args) {
        PalindromeLinkedList solution = new PalindromeLinkedList();
        // Input array
        int[] arr = {1, 2, 3, 1};

        // Convert array to linked list
        MyListNode dummy = new MyListNode(-1);
        MyListNode current = dummy;
        for (int num : arr) {
            current.next = new MyListNode(num);
            current = current.next;
        }
        MyListNode head = dummy.next;

        // Test the function
        System.out.println(solution.isPalindrome(head)); // Output: true
    }
}

File: python/PalindromeLinked List.py
class MyListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class PalindromeLinkedList:
    def is_palindrome(self, head):
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        if fast:  # odd nodes: let right half smaller
            slow = slow.next

        slow = self.reverse(slow)
        fast = head

        while slow:
            if fast.val != slow.val:
                return False
            fast = fast.next
            slow = slow.next

        return True

    def reverse(self, head):
        prev = None
        while head:
            next_node = head.next
            head.next = prev
            prev = head
            head = next_node
        return prev


if __name__ == "__main__":
    solution = PalindromeLinkedList()

    # Example usage
    # Input: head = [1,2,2,1]
    head = MyListNode(1)
    head.next = MyListNode(2)
    head.next.next = MyListNode(2)
    head.next.next.next = MyListNode(1)

    print(solution.is_palindrome(head))  # Output: True
